### DISCUSSION TITLE
Possibility to specify the virtual environment to use

### DISCUSSION THREAD

**Discussion by cerna** [pdm-project/pdm#397] [Q&A] [answered]
Hello,
I noticed that in file `.pdm.toml` there is specified the exact path of Python interpreter to use. **I am thinking if there could be also an exact path to virtual environment to use for build and install?** As there already can be `use_venv = true` configuration.

My use case is a bit particular:
I have a project (repository) assembled as a workspace. (In other words the modules are acting as if were up to a point independent.) The project is build with CMake and written in C, C++, Bash and Python. The Python part is comprised from extension modules for CPython wrapping calls to other C parts and pure Python modules building on top of the extensions. So everything has to be build in order and have dependencies on previous builds.

I would like for CMake to copy the Python `.py` files to BINARY tree (CMake has strong distinction between _SOURCE_ and _BINARY_ trees), template the _pyproject.toml_ there (set version, description etc) and build and install the package using specific virtual environment so all packages (distributions really) are using the same virtual environment and thus have the dependencies already present. You would be then able to use these "installs" after activating the virtual environment.

I don't have strong opinion if it should be virtual environment or the `__pypackages__`, but the later would probably fail on each package being in a different folder.

**Comment by frostming** [comment, marked as answer]
If I understand your request correctly, it is doable with PDM:

https://pdm.fming.dev/usage/project/#working-with-a-virtualenv

**Reply by cerna** [reply]
Thank you, I didn't realize the implications of
>Besides, when use-venv is on and the interpreter path given is a venv-like path, PDM will reuse that venv directory as well.

and then come upon it by accident. **It's great it is not just a temporary by-product, but a feature!**

Also, when I do the `pdm install`, it will install the package in __development/editable__ mode and create these `*.egg-link` files back to the code. You can then see the package with path when running `pip list`. However, when I try to do `pdm install` to the same virtual environment in other Python project which depends on the first one, it cannot resolve the dependency. But if I install this package as built wheel via `pip install dist/<name>.whl`, then everything works, i.e. __pip__ doesn't care about the dependency being installed in developer mode. Can I somehow get the same resolution with PDM?

(BTW, Poetry is the same, even though it is creating a `*.dist-info` folder with metadata when running the `poetry install` command.

**Reply by frostming** [reply]
Can you clarify your case with detailed steps to reproduce?

**Reply by cerna** [reply]
I created a simple example in [this](https://github.com/cerna/python_pdm_example/) repository. The detailed steps are in README.

**Reply by frostming** [reply]
Because PDM requires locked candidates before installation. For locking, the package must actually exist on whatever index you choose(default: PyPI)

But pip install only performs installation and checks against the currently installed packages. So it can be installed successfully. This is a difference in philosophy and I am afraid PDM can't achieve the same as pip.
